Better error responses

aldeed opened the pull request:
Adds ability for `data` argument of `sendResult` to be an `Error`, and catches errors in json-routes and automatically responds.

End goal is to be able to do this:

```js
Meteor.methods({
  foo: function (obj) {
    check(obj, objSchema);
  }
});
```

And get full schema validation info in the response object. I think this will also automatically work for POST/PATCH on collections that have a schema attached via collection2 package.

Let me know your thoughts. Thanks!

aldeed commented:
Probably should also add `prettyPrintResponses` option to method/publish options for fine-grained control/override.

stubailo commented on packages/json-routes/README.md:
```diff
@@ -36,13 +36,13 @@ Return data fom a route.
-- `code` - the status code to send. `200` for OK, `500` for internal error, etc.
-- `data` - the data you want to send back, will be sent as serialized JSON with
-content type `application/json`.
+- `code` - the status code to send. `200` for OK, `500` for internal error, etc. If the `data` argument is an `Error` type, this is overwritten based on the error.
```
In DDP, I think the best practice is to send string error codes. I'd consider it a pretty bad downgrade to encourage people to send numerical codes instead to comply with HTTP.

aldeed replied:
Does setStatusCode work with a string? There are a lot of RESTful best practices that involve sending specific numeric error codes, but I agree that it should support sending string codes.

stubailo replied:
I think I wasn't clear - the `code` option should still be a number, but it shouldn't automatically come from the error, since that would encourage people to use numerical DDP errors.

stubailo commented on packages/json-routes/README.md:
````diff
@@ -51,6 +51,18 @@ Set the headers returned by `JsonRoutes.sendResult`. Default value is:
 }
 ```
 
+### JsonRoutes.config(configObj)
````
I'm starting to reconsider this global options thing - it might not be great if you are using multiple packages that use JsonRoutes. Any ideas for a better way to do this?

aldeed replied:
Yeah, we already ran into an issue of needing to set a specific header for one route, which we did by calling `JsonRoutes.setResponseHeaders` in the route, but that's not ideal since it's global. I haven't given it enough thought to recommend a solution yet, though.

stubailo replied:
Maybe we could have a constructor for some context object that lets you share settings? Like:

```js
var CORSRoutes = new JsonRoutes.Context({
  headers: xxx,
  prefix: "/api"
});

CORSRoutes.add("name", header);
```

aldeed replied:
I like that.

There may be cases where headers need to be added dynamically, too, though, so `this.addHeader()` might be needed in route and method/publish scope. Similarly `this.setStatusCode()` for overriding the default 200/400/500. They can be no-op for DDP.

stubailo replied:
Well JSONRoutes supports those already through the response object, right? But yeah we might need something for Methods/Publications.

kahmali replied:
One thing you could consider is to export `JsonRoutes` as a class, and use its constructor to pass in the configuration options. That way the `JsonRoutes.add` method doesn't become nearly useless (since any routes added with `JsonRoutes` wouldn't be configurable). I've adopted this style in Restivus to support [basic versioning in the API](](https://github.com/kahmali/meteor-restivus/issues/64)). It's borrowed from Grape (a REST framework for Rack/Rails). The API would look very similar to what you have above:
```javascript
var CorsRoutes = new JsonRoutes({
  defaultHeaders: xxx,
  prefix: "/api"
});

CorsRoutes.add("name", header);
```

I also agree that there should be a way to set headers and the status code from the Methods/Publications.

stubailo commented on packages/json-routes/json-routes.js:
```diff
@@ -55,15 +55,60 @@ var setHeaders = function (res) {
+
+  // Convert `Error` objects to JSON representations
+  if (json) {
+    if (json instanceof Meteor.Error) {
```
I'd factor this out into a function. We will need that later anyway.

aldeed replied:
Done

stubailo commented on packages/json-routes/README.md:
```diff
@@ -36,13 +36,13 @@ Return data fom a route.
-content type `application/json`.
+- `code` - the status code to send. `200` for OK, `500` for internal error, etc. If the `data` argument is an `Error` type, this is overwritten based on the error.
+- `data` - the data you want to send back. This is serialized to JSON with
+content type `application/json`. If `undefined`, there will be no response body. If an `Error` type, a JSON representation of the error details will be sent.
```
Would it be better to let people throw errors instead of returning them?

aldeed replied:
Possibly. I remember thinking that briefly. You catch errors in the simple:rest and password packages anyway, so it makes sense to move that error catching logic down a level into json-routes.

stubailo replied:
Maybe. Maybe there is a need for an intermediate package that does stuff like catching errors and authentication, and this package can remain more simple?

stubailo commented on packages/json-routes/json-routes.js:
```diff
@@ -55,15 +55,60 @@ var setHeaders = function (res) {
+  // Set response body
   if (json !== undefined) {
-    var spacer = process.env.NODE_ENV === 'development' ? 2 : null;
+    var shouldPrettyPrint = (config.prettyPrintResponses || process.env.NODE_ENV === 'development');
```
Does this mean that people will sometimes want to pretty print in production? Why?

aldeed replied:
We plan to pretty print everything in production. We've compiled our best practices from a number of articles on the subject, but this is one of the better ones: http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#pretty-print-gzip

stubailo replied:
SGTM

stubailo commented:
Mostly I disagree with the HTTP error codes: https://github.com/meteor/meteor/blob/devel/packages/ddp/DDP.md#errors

In general, from what I've read on the internet, some people agree that HTTP error codes don't make sense, and shouldn't be used in a fine-grained way. Basically, you could just have three codes: 200 for success, 400 for invalid request, or 500 for an internal error.

It might be a good idea to let people set the HTTP error code anyway, but in that case I would like to decouple it from the `Meteor.Error` first argument.

aldeed commented:
My idea with checking the Error for error code was to handle all the existing code that puts numeric codes there. But in general I agree with your disagreement. :) And maybe it's fine to remove that.

Regarding having just three codes, one of our best practice ideas is to use 201 Created for resource creation. It's a pretty common standard and some clients look for it. I was thinking about proposing that the insert mutation restpoints should return 201 Created.

Some of the 400 codes are useful, too, but I can see where we might want to simply rely on a more understandable error string in the response body, which is also a bit cleaner to check in the response over both ddp and http.

stubailo commented:
Which clients look for a 201 code on creation success? I haven't heard of that before.

But yeah I'm not against letting people send whatever code they want, I just want that to be configurable separately from the DDP error string.

aldeed commented:
I'm not sure if clients look for it in practice, but the theory I read is that 201 Created should be accompanied by Location header referencing the resource so that it could be displayed right away.

stubailo commented:
Interesting. If it's truly a best practice we should do it but I'm not 100% sure it's that common? Maybe we can be ahead of the curve here.

stubailo commented:
@aldeed let's split out the error code and the pretty printing config into two PRs? I think we should do the context thing.

kahmali commented:
@stubailo I've come across the standard `201 Created` in multiple different resources, most important of which are [RFC 2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html) and something that I only discovered moments ago -- [RFC 7231](http://tools.ietf.org/html/rfc7231#section-6.3.2). Apparently, [RFC 2616 is being replaced](https://www.mnot.net/blog/2014/06/07/rfc2616_is_dead). Restivus conforms to this standard and returns a `201` on a successful POST to a collection route. According to best practices, POST should be used when creating resources, and the expected response code should be a `201`.

aldeed commented:
@stubailo, I removed the pretty printing / config stuff. I also updated json-routes to catch errors and automatically respond, and I removed the try/catches from the higher level packages. I think it makes a lot of sense this way.

I also have it checking for `jsonResponse` and `statusCode` properties added to the thrown error, which are used if present. This should allow me to automatically throw errors with full validation details from simple-schema. If you'd like to handle this some different way, I'm open to suggestions.

aldeed commented:
Oh, also added `setStatusCode()` in method context.

aldeed commented:
This also fixes https://github.com/stubailo/meteor-rest/issues/17

stubailo commented:
@aldeed it would be great if you could re-open this as a branch on this repo so that I can push to the branch directly. Also, rather than merging, it would be a good idea to rebase in order to not have a bunch of merge commits.

aldeed commented:
See https://github.com/stubailo/meteor-rest/pull/26
